### ISSUE TITLE
-Wthread-safety-reference doesn't warn when a reference to a guarded member is returned

### ISSUE THREAD

**Issue by pirama-arumuga-nainar** [llvm/llvm-project#66152]
`-Wthread-safety-reference` warns when a guarded member is returned by reference.

```
$ cat thread-safety.cpp 
#include <mutex>
#include <string>

extern void bar(std::string &s);

class C {
  private:
    std::mutex lock;
    std::string s __attribute__((guarded_by(lock)));

  public:
    std::string &foo() {
      bar(s); // warning emitted
      return s; // no warning
    }
};

$ clang++ -Wthread-safety -c thread-safety.cpp  -stdlib=libc++ -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS 
thread-safety.cpp:13:11: warning: passing variable 's' by reference requires holding mutex 'lock' [-Wthread-safety-reference]
      bar(s);
          ^
1 warning generated.
```

While returning by reference is potentially a bug, it is not invalid.  Adding a `REQUIRES` annotation to this function is valid.

**Comment by pirama-arumuga-nainar** [comment]
Covered by `-Wthread-safety-reference-return`.
